### PR TITLE
BUG: Breaking change to offsets.pyx - quarter offset classes

### PR DESCRIPTION
**Pull request by Pawel-Kranzberg**
Set default months for `BQuarterEnd`, `BQuarterBegin`, `QuarterEnd`, `QuarterBegin` classes in order; rename `startingMonth` parameter to `month` for those classes.
Addresses #8435 and #5307

- [ ] closes #8435 #5307
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
pls just one PR - just push to the same one if changing

**Comment by Pawel-Kranzberg** [comment]
> pls just one PR - just push to the same one if changing

But this one has different scope from #39887 .

**Comment by jreback** [comment]
ok that's fine

you always need a test

**Review by jreback — CHANGES_REQUESTED** [review]
you would need to change / add tests to show what is changing. and have the tests pass

**Comment by jbrockmendel** [comment]
does this handle #5307?

**Comment by Pawel-Kranzberg** [comment]
> does this handle #5307?

Not at present. I'd like to cover it as well - please let me know if that's OK.

**Comment by Pawel-Kranzberg** [comment]
> does this handle #5307?

It does now.

**Review by jreback — CHANGES_REQUESTED** [review]
so you are adding a new parameter 'month', so now we have the confusing 'startingMonth' and 'month'. I would rather deprecate 'startingMonth', then add new; not even sure 'month' is a good name here. can you write a sample whatsnew note indicate what the change is here from a user POV.

File: pandas/_libs/tslibs/offsets.pyx [hunk 1]
```diff
@@ -1840,7 +1840,7 @@ cdef class YearOffset(SingleConstructorOffset):
     """
     _attributes = tuple(["n", "normalize", "month"])
 
-    # _default_month: int  # FIXME: python annotation here breaks things
+    # FIXME: python annotation here breaks things
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx:1843]: do we know what thi refers?

File: pandas/_libs/tslibs/offsets.pyx [hunk 2]
```diff
@@ -1986,52 +1986,90 @@ cdef class YearBegin(YearOffset):
...
+    n : int, default 1
+        Number of quarters to offset.
+    normalize : bool, default False
+        If true, the time component of the resulting date-time is converted
+        to 00:00:00, i.e. midnight (the start, not the end of date-time).
+    month : int
+        The calendar month number (1 for January) of the beginning or ending
+        month in a custom-defined quarter to be used as an offset.
+
+        .. versionadded:: 1.X.X
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx:2003]: 1.3

File: pandas/_libs/tslibs/offsets.pyx [hunk 3]
```diff
@@ -1986,52 +1986,90 @@ cdef class YearBegin(YearOffset):
...
+    _attributes = tuple(["n", "normalize", "month"])
     # TODO: Consider combining QuarterOffset and YearOffset __init__ at some
-    #       point.  Also apply_index, is_on_offset, rule_code if
-    #       startingMonth vs month attr names are resolved
+    #       point. Also apply_index, is_on_offset, rule_code.
 
-    # FIXME: python annotations here breaks things
-    # _default_starting_month: int
-    # _from_name_starting_month: int
+    # FIXME: python annotation here breaks things
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx:2015]: can you be more specific

File: pandas/_libs/tslibs/offsets.pyx [hunk 4]
```diff
@@ -1986,52 +1986,90 @@ cdef class YearBegin(YearOffset):
...
 
     cdef readonly:
-        int startingMonth
+        int month
+        int startingMonth  # Backwards compatibility
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx:2019]: why is this needed? (e.g. can you simply assign it to month when needed?)

File: pandas/_libs/tslibs/offsets.pyx [hunk 5]
```diff
@@ -1986,52 +1986,90 @@ cdef class YearBegin(YearOffset):
...
 
-    def __init__(self, n=1, normalize=False, startingMonth=None):
+    def __init__(self, n=1, normalize=False, month=None,
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx:2021]: can you type

File: pandas/_libs/tslibs/offsets.pyx [hunk 6]
```diff
@@ -1986,52 +1986,90 @@ cdef class YearBegin(YearOffset):
...
+                 *, startingMonth=None):  # Backwards compatibility
         BaseOffset.__init__(self, n, normalize)
 
-        if startingMonth is None:
-            startingMonth = self._default_starting_month
-        self.startingMonth = startingMonth
+        # Backwards compatibility
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx:2025]: add the issue number here as a comment

File: pandas/_libs/tslibs/offsets.pyx [hunk 7]
```diff
@@ -1986,52 +1986,90 @@ cdef class YearBegin(YearOffset):
...
+        if month and startingMonth and month != startingMonth:
+            raise TypeError("Offset received both month and startingMonth")
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx:2027]: is this line tested?

File: pandas/_libs/tslibs/offsets.pyx [hunk 8]
```diff
@@ -1986,52 +1986,90 @@ cdef class YearBegin(YearOffset):
...
+        elif month is None and startingMonth is not None:
+            warnings.warn(
+                "startingMonth is deprecated, use month instead.",
+                DeprecationWarning,
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx:2031]: FutureWarning

File: pandas/_libs/tslibs/offsets.pyx [hunk 9]
```diff
@@ -1986,52 +1986,90 @@ cdef class YearBegin(YearOffset):
...
+        self.month = month
+
+        if month < 1 or month > 12:
+            raise ValueError("Month must go from 1 to 12.")
 
     cpdef __setstate__(self, state):
-        self.startingMonth = state.pop("startingMonth")
+        try:
+            self.month = state.pop("month")
+        except:
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx:2045]: can you be more specifi on the except

File: pandas/_libs/tslibs/offsets.pyx [hunk 10]
```diff
@@ -2092,26 +2148,44 @@ cdef class BQuarterBegin(QuarterOffset):
...
+    ----------
+    n : int, default 1
+        Number of business quarters to offset.
+    normalize : bool, default False
+        If true, the time component of the resulting date-time is converted
+        to 00:00:00, i.e. midnight (the start, not the end of date-time).
+    month : int, default 1
+        The calendar month number (1 for January) of the first month
+        in a custom-defined business quarter to be used as an offset.
+
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx:2165]: 1.3 for here and others

File: pandas/tests/tseries/offsets/test_offsets.py [hunk 11]
```diff
@@ -870,3 +873,17 @@ def test_dateoffset_immutable(attribute):
...
+@pytest.mark.parametrize(
+    "offset_type",
+    [
+        "QuarterBegin",
+        "QuarterEnd",
+        "BQuarterBegin",
+        "BQuarterEnd",
+    ],
+)
+def test_startingMonth_deprecation_warning(offset_type):
```
Comment (jreback) [inline pandas/tests/tseries/offsets/test_offsets.py:887]: FutureWarning and add the issue number as a comment

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a sub-section in the whatsnew 1.3 api breaking changes to indicate what a user would except (aside from the deprection)

File: pandas/_libs/tslibs/offsets.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/offsets.pyx]: putting a python-style annotation `_default_month: int` here led to (i think) compile-time errors

File: pandas/_libs/tslibs/offsets.pyx (hunk 10 above)
Comment (Pawel-Kranzberg) [reply pandas/_libs/tslibs/offsets.pyx:2163]: Done.

File: pandas/tests/tseries/offsets/test_offsets.py (hunk 11 above)
Comment (Pawel-Kranzberg) [reply pandas/tests/tseries/offsets/test_offsets.py]: Done.

File: pandas/_libs/tslibs/offsets.pyx (hunk 8 above)
Comment (Pawel-Kranzberg) [reply pandas/_libs/tslibs/offsets.pyx]: Done.

File: pandas/_libs/tslibs/offsets.pyx (hunk 7 above)
Comment (Pawel-Kranzberg) [reply pandas/_libs/tslibs/offsets.pyx:2024]: Thank you, I missed that one. Now it is tested.

File: pandas/_libs/tslibs/offsets.pyx (hunk 5 above)
Comment (Pawel-Kranzberg) [reply pandas/_libs/tslibs/offsets.pyx]: Done.

File: pandas/_libs/tslibs/offsets.pyx (hunk 6 above)
Comment (Pawel-Kranzberg) [reply pandas/_libs/tslibs/offsets.pyx]: Done.

File: pandas/_libs/tslibs/offsets.pyx (hunk 2 above)
Comment (Pawel-Kranzberg) [reply pandas/_libs/tslibs/offsets.pyx]: Done.

File: pandas/_libs/tslibs/offsets.pyx (hunk 1 above)
Comment (Pawel-Kranzberg) [reply pandas/_libs/tslibs/offsets.pyx]: Yes, that's right. I rolled back this change for now, as this part doesn't concern quarters. If OK, I hope to refactor/combine the  QuarterOffset and YearOffset and deal with it at some point.

File: pandas/_libs/tslibs/offsets.pyx (hunk 4 above)
Comment (Pawel-Kranzberg) [reply pandas/_libs/tslibs/offsets.pyx]: You are right. Now I left just the month declaration.

File: pandas/_libs/tslibs/offsets.pyx (hunk 9 above)
Comment (Pawel-Kranzberg) [reply pandas/_libs/tslibs/offsets.pyx:2042]: I added a longer comment there.

File: pandas/_libs/tslibs/offsets.pyx (hunk 3 above)
Comment (Pawel-Kranzberg) [reply pandas/_libs/tslibs/offsets.pyx]: I removed that comment, as other variables are now sufficiently typed - month in particular. Non-ints as default values would be captured by month.

**Comment by Pawel-Kranzberg** [comment]
> can you add a sub-section in the whatsnew 1.3 api breaking changes to indicate what a user would except (aside from the deprection)

Description drafted and added.

**Comment by Pawel-Kranzberg** [comment]
> so you are adding a new parameter 'month', so now we have the confusing 'startingMonth' and 'month'. I would rather deprecate 'startingMonth', then add new; not even sure 'month' is a good name here. can you write a sample whatsnew note indicate what the change is here from a user POV.

BTW, maybe `anchor` would be a better replacement name?

File: doc/source/whatsnew/v1.3.0.rst [hunk 12]
```diff
@@ -458,6 +458,36 @@ with a :class:`MultiIndex` in the result. This can lead to a perceived duplicati
...
 
+.. _whatsnew_130.notable_bug_fixes.quarterbegin_default_anchor:
+
+Changed default periods in :class:`~pandas.tseries.offsets.QuarterBegin` and :class:`~pandas.tseries.offsets.BQuarterBegin`
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Quarter-based time offset classes :class:`~pandas.tseries.offsets.QuarterEnd`,
+:class:`~pandas.tseries.offsets.QuarterBegin`, :class:`~pandas.tseries.offsets.BQuarterEnd`,
+:class:`~pandas.tseries.offsets.BQuarterBegin` now use ``month`` instead of ``startingMonth``
+as a time anchor parameter (:issue:`5307`). Its default value for :class:`~pandas.tseries.offsets.QuarterBegin`
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.3.0.rst:469]: "Its" -> "The"?

File: pandas/_libs/tslibs/offsets.pyx [hunk 13]
```diff
@@ -1986,52 +1986,88 @@ cdef class YearBegin(YearOffset):
...
+    # TODO: Consider combining QuarterOffset and YearOffset __init__ at some
+    #       point. Also apply_index, is_on_offset, rule_code.
 
     cdef readonly:
-        int startingMonth
+        int month
 
-    def __init__(self, n=1, normalize=False, startingMonth=None):
+    def __init__(self, n: int=1, normalize: bool=False, month: int=None,
+                 *, startingMonth: int=None):  # GH#5307 backwards compatibility
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/offsets.pyx:2019]: comment on new line `# GH#5307 startingMonth retained for backwards compatibility`

File: pandas/_libs/tslibs/offsets.pyx [hunk 14]
```diff
@@ -1986,52 +1986,88 @@ cdef class YearBegin(YearOffset):
...
+    _attributes = tuple(["n", "normalize", "month"])
+    # TODO: Consider combining QuarterOffset and YearOffset __init__ at some
+    #       point. Also apply_index, is_on_offset, rule_code.
 
     cdef readonly:
-        int startingMonth
+        int month
 
-    def __init__(self, n=1, normalize=False, startingMonth=None):
+    def __init__(self, n: int=1, normalize: bool=False, month: int=None,
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/offsets.pyx:2018]: should we try to detect if month is passed as a non-keyword, in which case the behavior may be silently changing?

(if backward compat werent an issue, I'd be OK with making everything other than `n` and `normalize` keyword-only in all these classes)

File: pandas/_libs/tslibs/offsets.pyx [hunk 15]
```diff
@@ -1986,52 +1986,88 @@ cdef class YearBegin(YearOffset):
...
 
-        if startingMonth is None:
-            startingMonth = self._default_starting_month
-        self.startingMonth = startingMonth
+        # GH#5307 backwards compatibility
+        if month and startingMonth and month != startingMonth:
+            raise TypeError("Offset received both month and startingMonth")
+        elif month is None and startingMonth is not None:
+            warnings.warn(
+                "startingMonth is deprecated, use month instead.",
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/offsets.pyx:2027]: should this explain that the meaning has changed?  or point to a link?

File: pandas/_libs/tslibs/offsets.pyx [hunk 16]
```diff
@@ -1986,52 +1986,88 @@ cdef class YearBegin(YearOffset):
...
+                FutureWarning,
+                stacklevel=2
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/offsets.pyx:2029]: trailing comma, pretend we could run `black` on this file

File: pandas/_libs/tslibs/offsets.pyx (hunk 9 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/offsets.pyx:2042]: a) im pretty sure jreback meant not to use a bare "except:", but to specifically catch `KeyError`

b) rule of thumb: pls let the person who asked the question hit the "resolve conversation" button; something might be clear to you but not to them

File: doc/source/whatsnew/v1.3.0.rst (hunk 12 above)
Comment (Pawel-Kranzberg) [reply doc/source/whatsnew/v1.3.0.rst]: Good catch, fixed.

File: pandas/_libs/tslibs/offsets.pyx (hunk 13 above)
Comment (Pawel-Kranzberg) [reply pandas/_libs/tslibs/offsets.pyx]: Comment added.

File: pandas/_libs/tslibs/offsets.pyx (hunk 16 above)
Comment (Pawel-Kranzberg) [reply pandas/_libs/tslibs/offsets.pyx]: Fixed, hopefully.

File: doc/source/whatsnew/v1.3.0.rst [hunk 17]
```diff
@@ -458,6 +458,36 @@ with a :class:`MultiIndex` in the result. This can lead to a perceived duplicati
     df.groupby('label1').rolling(1).sum()
 
 
+.. _whatsnew_130.notable_bug_fixes.quarterbegin_default_anchor:
+
+Changed default periods in :class:`~pandas.tseries.offsets.QuarterBegin` and :class:`~pandas.tseries.offsets.BQuarterBegin`
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:463]: this needs to be the same length as the title, it should error if not

File: doc/source/whatsnew/v1.3.0.rst [hunk 18]
```diff
@@ -458,6 +458,36 @@ with a :class:`MultiIndex` in the result. This can lead to a perceived duplicati
...
+.. code-block:: ipython
+
+   In [1]: pd.Timestamp(2014, 10, 10) + pd.tseries.offsets.QuarterBegin()
+   Out[2]: Timestamp('2014-12-01 00:00:00')
+   In [3]: pd.Timestamp(2014, 10, 10) + pd.tseries.offsets.BQuarterBegin()
+   Out[4]: Timestamp('2014-12-01 00:00:00')
+
+*pandas 1.3.0*
+
+.. code-block:: ipython
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:484]: don't use a code-block, rather an ipython block for the new behavior

File: doc/source/whatsnew/v1.3.0.rst [hunk 19]
```diff
@@ -458,6 +458,36 @@ with a :class:`MultiIndex` in the result. This can lead to a perceived duplicati
     df.groupby('label1').rolling(1).sum()
 
 
+.. _whatsnew_130.notable_bug_fixes.quarterbegin_default_anchor:
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:461]: you need to say deprecation about the `startingMonth` as its mentioned below as well. I would prefer making these 2 distinct and unrelated changes (which they are). so don't mention deprecation at all here.

File: doc/source/whatsnew/v1.3.0.rst [hunk 20]
```diff
@@ -625,6 +656,7 @@ Datetimelike
 - Bug in :meth:`Timedelta.round`, :meth:`Timedelta.floor`, :meth:`Timedelta.ceil` for values near the implementation bounds of :class:`Timedelta` (:issue:`38964`)
 - Bug in :func:`date_range` incorrectly creating :class:`DatetimeIndex` containing ``NaT`` instead of raising ``OutOfBoundsDatetime`` in corner cases (:issue:`24124`)
 - Bug in :func:`infer_freq` incorrectly fails to infer 'H' frequency of :class:`DatetimeIndex` if the latter has a timezone and crosses DST boundaries (:issue:`39556`)
+- Bug in offsets :class:`QuarterBegin` and :class:`BQuarterBegin` returning days that are not conventional quarter beginnings (:issue:`8435`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:659]: isn't this a duplicate of the above section?

File: pandas/_libs/tslibs/offsets.pyx [hunk 21]
```diff
@@ -1986,52 +1986,88 @@ cdef class YearBegin(YearOffset):
...
+    def __init__(self, n: int=1, normalize: bool=False, month: int=None,
+                 *, startingMonth: int=None):
+        # GH#5307 startingMonth retained for backwards compatibility
         BaseOffset.__init__(self, n, normalize)
 
-        if startingMonth is None:
-            startingMonth = self._default_starting_month
-        self.startingMonth = startingMonth
+        # GH#5307 startingMonth retained for backwards compatibility
+        if month and startingMonth and month != startingMonth:
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx:2024]: don't you need to check not-none? e.g. 0 is value that would hit this too (which would fail but is misleading no?)

File: pandas/_libs/tslibs/offsets.pyx (hunk 9 above)
Comment (jreback) [reply pandas/_libs/tslibs/offsets.pyx:2042]: yeah this is non-obvious on the exception.

**Review by jreback — CHANGES_REQUESTED** [review]
@Pawel-Kranzberg can you split this change up. There is just too much going on here.

- doc-strings
- deprecation
- breaking change
-
I would *really* like to do the deprecation first and separate

**Comment by simonjayhawkins** [comment]
> I would _really_ like to do the deprecation first and separate

@Pawel-Kranzberg what's the status. close this and open separate PRs?

**Comment by Pawel-Kranzberg** [comment]
> > I would _really_ like to do the deprecation first and separate
> 
> @Pawel-Kranzberg what's the status. close this and open separate PRs?

Yes, I plan to do so next week.

**Comment by mroeschke** [comment]
Thanks for the PR, but it has appeared to have gone stale. Closing due to inactivity, but we would be happy to take a new PR working on the deprecation first.
